onnxruntime/core/providers/openvino/ov_stateful_patch_utils.cc

```diff
@@ -134,50 +134,113 @@ void MakeStateful(std::shared_ptr<ov::Model>& ov_model,
+      const auto& names = input.get_names();
+      const std::string input_name = input.get_any_name();
+
+      bool is_kv_input = false;
+      for (const auto& name : names) {
+        if (name.find("key_values") != std::string::npos ||
+            name.find("keys") != std::string::npos ||
+            name.find("values") != std::string::npos) {
+          key_value_input_names.push_back(name);
+          is_kv_input = true;
+          break;
+        }
+      }
+
+      if (!is_kv_input) {
+        not_kv_inputs.push_back(input_name);
+      }
+    }
+
+    return std::make_pair(key_value_input_names, not_kv_inputs);
+  }
+
+  std::set<std::string> found_kv_inputs;
+
   for (const ov::Output<ov::Node>& input : model->inputs()) {
     auto& names = input.get_names();
 
     bool found = false;
-    for (auto& name : names) {
-      if (name.find("key_values") != std::string::npos) {
-        key_value_input_names.push_back(name);
-        found = true;
-        break;
-      } else if (name.find("keys") != std::string::npos) {
-        key_value_input_names.push_back(name);
-        found = true;
-        break;
-      } else if (name.find("values") != std::string::npos) {
-        key_value_input_names.push_back(name);
-        found = true;
-        break;
+
+    // Check if any input name contains the extracted patterns
+    for (const auto& name : names) {
+      for (const auto& pattern : patterns) {
+        if (name.find(pattern) != std::string::npos){
+          key_value_input_names.push_back(name);
+          found = true;
+          break;
+        }
       }
+      if (found) break;
     }
 
     if (!found) {
       not_kv_inputs.push_back(input.get_any_name());
     }
   }
 
-  std::vector<std::string> key_value_output_names;
-  for (const ov::Output<ov::Node>& output : model->outputs()) {
-    auto& names = output.get_names();
-    for (auto& name : names) {
-      if (name.find("present") != std::string::npos) {
-        key_value_output_names.push_back(name);
-        break;
-      }
-    }
-  }
+  return std::make_pair(key_value_input_names, not_kv_inputs);
+}
 
+// Updated PatchStatefulDecoder function
+void PatchStatefulDecoder(std::shared_ptr<ov::Model> model) {
+  // Use the dynamic pattern-based extraction logic
+  auto [key_value_output_names, extracted_patterns] = ExtractKVPatternsFromOutputs(model);
+  auto [key_value_input_names, not_kv_inputs] = ExtractInputKVTensors(model, extracted_patterns);
+
+  std::cout << key_value_input_names.size() << ";" << key_value_output_names.size() << std::endl;
   if (key_value_input_names.empty() || key_value_output_names.empty()) {
     std::cout << "no key_value_input_names or key_value_output_names found" << std::endl;
     return;
   }
 
+  if (key_value_input_names.size() != key_value_output_names.size()) {
+    std::cout << "found different sizes btween key_value_input_names and key_value_output_names, they couldn't be paired" << std::endl;
```

Comment:
I think that this one should be a runtime exception of some sort. I don't think we'd ever want to hit this state, return, and have the rest of the stateful flow continue on.